File: data_structures/trees/binary_tree/python/binary_tree_create_inorder_postorder.py
from typing import List, Union, TypeVar, Final
from data_structures.trees.binary_tree.python.binary_tree_node import BinaryTreeNode

T = TypeVar("T")


def binary_tree_create_inorder_postorder_simple(inorder: List[T], postorder: List[T]) -> Union[BinaryTreeNode[T], None]:
    """
    Constructs binary tree from inorder and postorder traversals
      3
    9   20
      15  7
    """
    in_order_idx = len(inorder) - 1
    post_order_idx = len(postorder) - 1

    def build(inorder: List[T], postorder: List[T], root: BinaryTreeNode[T] = None) -> Union[BinaryTreeNode[T], None]:
        nonlocal post_order_idx
        nonlocal in_order_idx

        if post_order_idx < 0:
            return None

        node: Final[BinaryTreeNode[T]] = BinaryTreeNode(postorder[post_order_idx])
        post_order_idx -= 1

        if inorder[in_order_idx] != node.val:
            node.right = build(inorder, postorder, node)

        in_order_idx -= 1

        if root is None or inorder[in_order_idx] != root.val:
            node.left = build(inorder, postorder, root)

        return node

    return build(inorder, postorder, None)

File: data_structures/trees/binary_tree/typescript/binaryTreeCreateInOrderPostOrder.ts
import BinaryTreeNode from '@ds/trees/binary_tree/typescript/BinaryTreeNode';

export function binaryTreeCreateInOrderPostOrderSimple<T>(inOrder: T[], postOrder: T[]): BinaryTreeNode<T> {
    let inOrderIndex = inOrder.length - 1;
    let postOrderIndex = postOrder.length - 1;

    const build = (inOrder: T[], postOrder: T[], root: BinaryTreeNode<T>): BinaryTreeNode<T> => {
        if (postOrderIndex < 0) {
            return null;
        }

        const node = new BinaryTreeNode<T>(postOrder[postOrderIndex--]);

        if (inOrder[inOrderIndex] != node.val) {
            node.right = build(inOrder, postOrder, node);
        }

        inOrderIndex--;

        if (root === null || inOrder[inOrderIndex] !== root.val) {
            node.left = build(inOrder, postOrder, root);
        }

        return node;
    };

    return build(inOrder, postOrder, null);
}
